modules/internal/DefaultSparse.chpl

```diff
@@ -235,7 +235,114 @@ module DefaultSparse {
         add_help(ind);
       }
     }
-  
+
+    proc bulkAdd(inds: [] index(rank, idxType), sorted=false, noDuplicate=false){
+      var numAdded = inds.size; //maybe remove this var
+
+      //find individual insert points
+      //and eliminate duplicates between inds and dom
+      var indivInsertPts: [{0..#numAdded}] int;
+      var actualInsertPts: [{0..#numAdded}] int; //where to put in newdom
+
+      if !sorted then QuickSort(inds);
+
+      //eliminate duplicates --assumes sorted
+      if !noDuplicate {
+        //make sure lastInd != inds[inds.domain.low]
+        var lastInd = inds[inds.domain.low] + 1; 
+        for (i, p) in zip(inds, indivInsertPts)  {
+          if i == lastInd then p = -1;
+          else lastInd = i;
+        }
+      }
+
+      //verify sorted and no duplicates if not --fast
+      if boundsChecking {
+        VerifySort(inds, "bulkAdd: Data is not sorted, call the function with \
+            sorted=false");
+
+        //check duplicates assuming sorted
+        const indsStart = inds.domain.low;
+        const indsEnd = inds.domain.high;
+        var lastInd = inds[indsStart];
+        for i in indsStart+1..indsEnd {
+          if inds[i] == lastInd then halt("There are duplicates"); 
+        }
+
+        //couldn't use boundsCheck because of 1D indices -- they are not 1tuple
+        for i in inds do
+          if !parentDom.member(i) then
+            halt("Index out of bounds:", i);
+
+      }
+
+      for (i,p) in zip(inds, indivInsertPts) {
+        if p != -1 { //don't do anything if it's duplicate
+          const (found, insertPt) = find(i);
+          p = if found then -1 else insertPt; //mark as duplicate
+        }
+      }
+
+      //shift insert points for bulk addition
+      //previous indexes that are added will cause a shift in the next indexes
+      var actualAddCnt = 0;
+      for (ip, ap) in zip(indivInsertPts, actualInsertPts) {
```

Comment:
FYI I think this could be implemented with scan, but it might be better to do it this way for now.